### PR TITLE
[MERL-357] sitemaps without middleware response

### PR DESCRIPTION
**Pull request by matthewguywright**
## Description
Updated sitemaps to use sitemaps.xml and a query param of 'sitemap' to display sitemaps within the confines of no response body allowed in middleware.

Provide the GitHub issue(s) number for issue tracking purposes, use the following syntax:

- #MERL-357
-->

### CODE REVIEW

File: packages/next/src/server/getSitemapProps.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import { GetServerSidePropsContext } from 'next';
+import { createPagesSitemap, createRootSitemapIndex, handleSitemapPath } from '../middleware/sitemaps/createSitemaps.js';
+import { HandleSitemapRequestsConfig } from '../middleware/sitemaps/handleSitemapRequests.js';
+
+export async function getSitemapProps(
+  ctx: GetServerSidePropsContext,
+  config: any,
+) {
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (theodesp) [inline packages/next/src/server/getSitemapProps.tsx:9]: we might be very careful here with `ts-ignores` is there a particular rule that we don't follow?

File: packages/next/src/server/getSitemapProps.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    ctx.res.write(await response?.text());
+
+    ctx.res.end();
+  }
+
+  if (
+    urlParams.get('sitemap') &&
+    urlParams.get('sitemap') !== 'sitemap-faust-pages.xml'
+  ) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
```
Comment (theodesp) [inline packages/next/src/server/getSitemapProps.tsx:35]: We might want to see how to make it typesafe:

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-unsafe-argument.md

File: packages/next/src/middleware/sitemaps/createSitemaps.ts [hunk 3]
```diff
@@ -51,21 +52,54 @@ const parserConfig: Partial<X2jOptions> = {
...
+    const { pathname, origin } = new URL(req.url);
+    frontendUrl = origin;
+    wpSitemapUrl = `${trimSlashes(wpUrl)}/${trimSlashes(pathname)}`;
+  } else {
+    // get rootSitemapPath config param
+    // fetch sitemap from WP
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+    wpSitemapUrl = `${trimSlashes(wpUrl)}/${trimSlashes(
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
+      normalizedConfig.rootSitemapPath,
```
Comment (blakewilson) [inline packages/next/src/middleware/sitemaps/createSitemaps.ts:77]: I'm not sure why I missed this during our pair programming but we already have a config setting for the root sitemap path, it's called `sitemapIndexPath`. Let's use this instead as it's one less new option to introduce.

File: packages/next/src/middleware/sitemaps/createSitemaps.ts [hunk 4]
```diff
@@ -51,21 +52,54 @@ const parserConfig: Partial<X2jOptions> = {
...
+    } else {
+      sitemapFaustPagesUrl = `${trimSlashes(
+        frontendUrl,
+      )}/sitemap.xml?sitemap=${trimSlashes(FAUST_PAGES_PATHNAME)}`;
+    }
     sitemaps = [
       ...sitemaps,
-      { loc: `${trimSlashes(origin)}/${trimSlashes(FAUST_PAGES_PATHNAME)}` },
+      {
+        loc: sitemapFaustPagesUrl,
```
Comment (blakewilson) [inline packages/next/src/middleware/sitemaps/createSitemaps.ts:98]: We should run this through `encodeURI()`:

https://www.sitemaps.org/protocol.html#locdef

File: packages/next/src/server/getSitemapProps.tsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    const response = await handleSitemapPath(ctx.req, config, false);
+
+    ctx.res.setHeader('Content-Type', 'application/xml');
+
+    ctx.res.write(await response?.text());
```
Comment (blakewilson) [inline packages/next/src/server/getSitemapProps.tsx:47]: We'll want to check for 404s here if the sitemap does not exist on WordPress.

File: packages/next/src/middleware/sitemaps/handleSitemapRequests.ts [hunk 6]
```diff
@@ -102,6 +107,23 @@ export function validateConfig(
...
+      throw new Error('frontendUrl is required');
+    }
+
+    if (!isString(config?.frontendUrl)) {
+      throw new Error('frontendUrl must be a string');
+    }
+
+    try {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const url = new URL(config.frontendUrl);
```
Comment (theodesp) [inline packages/next/src/middleware/sitemaps/handleSitemapRequests.ts:121]: If we don't use the url you can omit it:
```javascript
new URL(config.frontendUrl)
```

File: packages/next/src/server/getSitemapProps.tsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+export async function getSitemapProps(
+  ctx: GetServerSidePropsContext,
+  config: NormalizedConfig,
+) {
+  // config validation with middleware flag
+  validateConfig(config, false);
+
+  // Normalize config if some optional values are missing
+  // eslint-disable-next-line prefer-object-spread
+  const normalizedConfig: NormalizedConfig = Object.assign(
```
Comment (theodesp) [inline packages/next/src/server/getSitemapProps.tsx:23]: use
```
Object.assign({},{...config, replaceUrls: true});
```
 to get rid of the `eslint-disable` rule

File: packages/next/src/middleware/sitemaps/createSitemaps.ts [hunk 8]
```diff
@@ -51,21 +52,54 @@ const parserConfig: Partial<X2jOptions> = {
...
+  let wpSitemapUrl = '';
+  let frontendUrl = '';
+  if (isMiddleware) {
+    const { pathname, origin } = new URL(req.url);
+    frontendUrl = origin;
+    wpSitemapUrl = `${trimSlashes(wpUrl)}/${trimSlashes(pathname)}`;
+  } else {
+    // get sitemapIndexPath config param
+    // fetch sitemap from WP
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
```
Comment (theodesp) [inline packages/next/src/middleware/sitemaps/createSitemaps.ts:74]: Both `eslint-disable` rules are not needed here

File: packages/next/src/middleware/sitemaps/handleSitemapRequests.ts [hunk 9]
```diff
@@ -36,11 +36,15 @@ export interface HandleSitemapRequestsConfig {
...
-   *
-   * @example /sitemap.xml, /wp-sitemap.xml, /sitemap_index.xml
+   * The sitemap path for server side
    */
   sitemapIndexPath: string;
+
+  /**
+   * The headless frontend URL
+   */
+  frontendUrl: string;
```
Comment (blakewilson) [inline packages/next/src/middleware/sitemaps/handleSitemapRequests.ts:46]: We may need to create a typescript interface `GetSItemapPropsConfig` that requires `frontendUrl`, and take it out of `HandleSitemapRequestsConfig` as that is not a valid property for the middleware.

Something like:

```ts
export interface GetSitemapPropsConfig extends HandleSitemapRequestsConfig {
  /**
   * The headless frontend URL
   */
  frontendUrl: string;
}
```

File: examples/next/getting-started/src/pages/sitemap.xml.tsx [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import "faust.config";
```
Comment (blakewilson) [inline examples/next/getting-started/src/pages/sitemap.xml.tsx:1]: We'll want to remove this file from the example and leave it up to the user to implement.

File: packages/next/src/middleware/sitemaps/createSitemaps.ts [hunk 11]
```diff
@@ -143,11 +177,23 @@ export async function createRootSitemapIndex(
...
     wpSitemaps.forEach((sitemap) => {
+      let sitemapUrl = '';
+      if (isMiddleware) {
+        sitemapUrl = sitemap.loc.replace(
+          trimSlashes(wpUrl),
+          trimSlashes(frontendUrl),
+        );
+      } else {
+        const url = new URL(sitemap.loc);
+        sitemapUrl = `${frontendUrl}/sitemap.xml?sitemap=${trimSlashes(
```
Comment (blakewilson) [inline packages/next/src/middleware/sitemaps/createSitemaps.ts:188]:
```suggestion
        sitemapUrl = `${trimSlashes(frontendUrl)}/sitemap.xml?sitemap=${trimSlashes(
```

**Comment by theodesp** [comment]
@matthewguywright awesome. Only the package-lock.json conflict left here.

**Comment by blakewilson** [comment]
One final thing we need is to add normalized config types for `getSitemapProps` and `handleSitemapRequests`. Currently the `handleSitemapRequests` type is broken as it requires the `frontendUrl`. I've made a quick PR into this branch to resolve this:

https://github.com/wpengine/faustjs/pull/982

**Comment by theodesp** [comment]
Few of the CI/CD steps are failing here.

**Comment by blakewilson** [comment]
@matthewguywright @theodesp These CI/CD steps should pass once #988 is merged into `canary`

**Review by theodesp — APPROVED** [review]
Awesome!

**Pull request merged**